### PR TITLE
Fix executor so a nested bundle call can succeed

### PR DESCRIPTION
**Pull request by xjunior**
Cleans up bundle environment so a nested bundle call can set its own environment variables

### CODE REVIEW

**Comment by xjunior** [comment]
@benlangfeld if we don't do this, when you run `./bin/cobra exec` it loads a bundle environment, and this environment is passed along overriding next bundle calls, if we do the way we were doing we're removing important environment variables. This is now the equivalent of actually cd'ing into the directory and running the command.

**Comment by benlangfeld** [comment]
> @benlangfeld if we don't do this, when you run `./bin/cobra exec` it loads a bundle environment, and this environment is passed along overriding next bundle calls, if we do the way we were doing we're removing important environment variables. This is now the equivalent of actually cd'ing into the directory and running the command.

Can you provide an example of something that previously didn't work and now does?

File: spec/cobra_commander/executor_spec.rb [hunk 1]
```diff
@@ -21,18 +21,9 @@
     expect(output.string).to match %r{===> node_manifest \(.*spec/fixtures/app/node_manifest\)}
   end
 
-  it "executes with clean environment" do
-    output = StringIO.new
-    CobraCommander::Executor.exec(umbrella.components, "env", output)
+  it "cleans BUNDLE environment so a nested bundle call can succeed" do
+    expect(Bundler).to receive(:with_original_env).and_yield
 
-    expect(output.string).to match %r{===> a \(.*spec/fixtures/app/components/a\)\n^$}
```
Comment (benlangfeld) [inline spec/cobra_commander/executor_spec.rb:28]: Why don't you need these assertions any more?

File: spec/cobra_commander/executor_spec.rb [hunk 2]
```diff
@@ -21,18 +21,9 @@
     expect(output.string).to match %r{===> node_manifest \(.*spec/fixtures/app/node_manifest\)}
   end
 
-  it "executes with clean environment" do
-    output = StringIO.new
-    CobraCommander::Executor.exec(umbrella.components, "env", output)
+  it "cleans BUNDLE environment so a nested bundle call can succeed" do
+    expect(Bundler).to receive(:with_original_env).and_yield
```
Comment (benlangfeld) [inline spec/cobra_commander/executor_spec.rb:25]: This isn't very valuable. Can we assert behaviour instead?

File: spec/cobra_commander/executor_spec.rb (hunk 1 above)
Comment (xjunior) [reply spec/cobra_commander/executor_spec.rb:28]: They were asserting the environment is clean, they're no longer valid as the environment could have something.

File: spec/cobra_commander/executor_spec.rb (hunk 2 above)
Comment (xjunior) [reply spec/cobra_commander/executor_spec.rb:25]: I'll think of a way of doing this

Comment (xjunior) [reply spec/cobra_commander/executor_spec.rb:25]: I'm trying to think of something valid to assert that the original bundler env var's are there but I can't think of anything that doesn't end-up testing exactly what `with_original_env` is designed for.

**Comment by xjunior** [comment]
> Can you provide an example of something that previously didn't work and now does?

A simple `./bin/schema` or `bundle exec rspec` was problematic.

**Comment by benlangfeld** [comment]
> > Can you provide an example of something that previously didn't work and now does?
> 
> A simple `./bin/schema` or `bundle exec rspec` was problematic.

Problematic in what way? How did it fail?

**Comment by xjunior** [comment]
@benlangfeld being lost:

```bash
➜  nitro git:(nova/periodic-snapshots) be cobra exec human_resources './bin/schema; bundle exec rspec'
===> human_resources (/Users/chjunior/workspace/power/nitro/components/human_resources)
>>> Setting up database
/System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/lib/ruby/2.6.0/bundler/spec_set.rb:91:in `block in materialize': Could not find nokogiri-1.10.4 in any of the sources (Bundler::GemNotFound)
	from /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/lib/ruby/2.6.0/bundler/spec_set.rb:85:in `map!'
	from /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/lib/ruby/2.6.0/bundler/spec_set.rb:85:in `materialize'
	from /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/lib/ruby/2.6.0/bundler/definition.rb:170:in `specs'
	from /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/lib/ruby/2.6.0/bundler/definition.rb:237:in `specs_for'
	from /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/lib/ruby/2.6.0/bundler/definition.rb:226:in `requested_specs'
	from /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/lib/ruby/2.6.0/bundler/runtime.rb:108:in `block in definition_method'
	from /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/lib/ruby/2.6.0/bundler/runtime.rb:20:in `setup'
	from /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/lib/ruby/2.6.0/bundler.rb:107:in `setup'
	from /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/lib/ruby/2.6.0/bundler/setup.rb:20:in `<top (required)>'
	from /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/lib/ruby/2.6.0/rubygems/core_ext/kernel_require.rb:54:in `require'
	from /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/lib/ruby/2.6.0/rubygems/core_ext/kernel_require.rb:54:in `require'
	from bin/rake:27:in `<main>'
bundler: command not found: rspec
Install missing gem executables with `bundle install`
```

**Comment by benlangfeld** [comment]
> @benlangfeld being lost:
> 
> ```shell
> ➜  nitro git:(nova/periodic-snapshots) be cobra exec human_resources './bin/schema; bundle exec rspec'
> ===> human_resources (/Users/chjunior/workspace/power/nitro/components/human_resources)
> >>> Setting up database
> /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/lib/ruby/2.6.0/bundler/spec_set.rb:91:in `block in materialize': Could not find nokogiri-1.10.4 in any of the sources (Bundler::GemNotFound)
> 	from /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/lib/ruby/2.6.0/bundler/spec_set.rb:85:in `map!'
> 	from /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/lib/ruby/2.6.0/bundler/spec_set.rb:85:in `materialize'
> 	from /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/lib/ruby/2.6.0/bundler/definition.rb:170:in `specs'
> 	from /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/lib/ruby/2.6.0/bundler/definition.rb:237:in `specs_for'
> 	from /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/lib/ruby/2.6.0/bundler/definition.rb:226:in `requested_specs'
> 	from /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/lib/ruby/2.6.0/bundler/runtime.rb:108:in `block in definition_method'
> 	from /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/lib/ruby/2.6.0/bundler/runtime.rb:20:in `setup'
> 	from /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/lib/ruby/2.6.0/bundler.rb:107:in `setup'
> 	from /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/lib/ruby/2.6.0/bundler/setup.rb:20:in `<top (required)>'
> 	from /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/lib/ruby/2.6.0/rubygems/core_ext/kernel_require.rb:54:in `require'
> 	from /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/lib/ruby/2.6.0/rubygems/core_ext/kernel_require.rb:54:in `require'
> 	from bin/rake:27:in `<main>'
> bundler: command not found: rspec
> Install missing gem executables with `bundle install`
> ```

Can we use that as the basis of a test?

**Comment by xjunior** [comment]
@benlangfeld the issue with this approach is that it also depends on the bundle config loaded in the CI environment

**Comment by xjunior** [comment]
@benlangfeld the usage of this method is pretty direct to what this is. We want the environment as it was found by bundler before it came into play, so when we execute `cd … && <command>` it behaves exactly as if you have typed that command directly.

**Comment by benlangfeld** [comment]
> @benlangfeld the usage of this method is pretty direct to what this is. We want the environment as it was found by bundler before it came into play, so when we execute `cd … && <command>` it behaves exactly as if you have typed that command directly.

I still think we need to assert that behaviour, else the test is completely pointless.

**Comment by xjunior** [comment]
@benlangfeld I think this covers it. Revert the fix and it will fail because bundler will be in a weird state.

**Pull request merged**